### ISSUE TITLE
InternalServerError: failed to lookup transaction or savepoint

### ISSUE THREAD

**Issue by jaclarke** [edgedb/edgedb#3120]
- EdgeDB Version: 1.0-rc.2+dev6135.d2021110220.ga97561df3.cv202111010000

Steps to Reproduce:

```python
import edgedb

def main():
    conn = edgedb.connect('_localdev')

    conn.query('start transaction')

    try:
        conn.query_single('rollback')
    except edgedb.InterfaceError:
        # ignore 'query cannot be executed with query_single() as it does not return any data' error
        pass

    conn.query('rollback')

    conn.close()

if __name__ == '__main__':
    main()
```
```
$ python test.py
Traceback (most recent call last):
  File "/home/james/dev/edgedb-python/test.py", line 19, in <module>
    main()
  File "/home/james/dev/edgedb-python/test.py", line 14, in main
    conn.query('rollback')
  File "/home/james/dev/edgedb-python/edgedb/blocking_con.py", line 346, in query
    return self._execute(
  File "/home/james/dev/edgedb-python/edgedb/blocking_con.py", line 325, in _execute
    raise e
  File "/home/james/dev/edgedb-python/edgedb/blocking_con.py", line 314, in _execute
    return self._get_protocol().sync_execute_anonymous(
  File "edgedb/protocol/blocking_proto.pyx", line 105, in edgedb.protocol.blocking_proto.BlockingIOProtocol.sync_execute_anonymous
    result, _headers = self._iter_coroutine(
  File "edgedb/protocol/blocking_proto.pyx", line 91, in edgedb.protocol.blocking_proto.BlockingIOProtocol._iter_coroutine
    coro.send(None)
  File "edgedb/protocol/protocol.pyx", line 617, in execute_anonymous
    codecs = await self._parse(
  File "edgedb/protocol/protocol.pyx", line 283, in _parse
    raise exc
edgedb.errors.InternalServerError: failed to lookup transaction or savepoint with id=35938955293271
```

**Comment by 1st1** [comment]
I've looked at this and I doubt we will fix it in 1.0.

The problem is that the compiler eagerly evaluates the compiled query against its transactional state and assumes that once the query is compiled it will be executed.

From the standpoint of the compiler, the first `conn.query_single('rollback')` would succeed so it assumes the transaction will be closed and kills its internal transactional state.

The way to fix this is to introduce some sort of phased synchronization between the IO process and the compiler: the compiler should build the state but not assume anything about it until the IO process actually confirms the execution. But I'm not sure how hard this would be to implement so I doubt we can (or should) fix this before 1.0. The risk of destabilizing the code base is too high imo.

**Comment by msullivan** [comment]
This is fixed
